File: js/src/Tested/NBodiesSimulation/Body.js
export class Body
{
    static _DT        = 0.1;
    static _SOFTENING = 1e-9;

    coordinate_x;
    coordinate_y;
    coordinate_z;

    velocity_x;
    velocity_y;
    velocity_z;

    mass;

    constructor()
    {
        this.coordinate_x = Math.random();
        this.coordinate_y = Math.random();
        this.coordinate_z = Math.random();

        this.mass = Math.random();
    }

    updateVelocity(another)
    {
        const dx = this.coordinate_x - another.coordinate_x;
        const dy = this.coordinate_y - another.coordinate_y;
        const dz = this.coordinate_z - another.coordinate_z;

        const distSq = Math.pow(dx, 2) + Math.pow(dy, 2) + Math.pow(dz, 2) + Body._SOFTENING;

        const invDist3 = Math.pow(1.0 / Math.sqrt(distSq), 3);

        this.velocity_x = dx * another.mass * invDist3 * Body._DT;
        this.velocity_y = dy * another.mass * invDist3 * Body._DT;
        this.velocity_z = dz * another.mass * invDist3 * Body._DT;
    }

    positionIntegration()
    {
        this.coordinate_x += this.velocity_x * Body._DT;
        this.coordinate_y += this.velocity_y * Body._DT;
        this.coordinate_z += this.velocity_z * Body._DT;
    }
}

File: c#/Benchmark/Logic/NBodySimulation/Body.cs
namespace Benchmark.Logic.NBodySimulation;

public class Body
{
    private const    double DT        = 0.01;
    private const    double SOFTENING = 1e-9;
    private readonly double _mass;
    private          double _coordinateX;
    private          double _coordinateY;
    private          double _coordinateZ;
    private          double _velocityX;
    private          double _velocityY;
    private          double _velocityZ;

    public Body()
    {
        this._coordinateX = Random.Shared.NextDouble();
        this._coordinateY = Random.Shared.NextDouble();
        this._coordinateZ = Random.Shared.NextDouble();
        this._velocityX   = Random.Shared.NextDouble();
        this._velocityY   = Random.Shared.NextDouble();
        this._velocityZ   = Random.Shared.NextDouble();
        this._mass        = Random.Shared.NextDouble();
    }

    public void UpdateVelocity(Body another)
    {
        double dx = this._coordinateX - another._coordinateX;
        double dy = this._coordinateY - another._coordinateY;
        double dz = this._coordinateZ - another._coordinateZ;

        double distSq   = Math.Pow(dx, 2) + Math.Pow(dy, 2) + Math.Pow(dz, 2) + Body.SOFTENING;
        double invDist3 = Math.Pow(1.0 / Math.Sqrt(distSq), 3);

        this._velocityX += dx * this._mass * invDist3 * Body.DT;
        this._velocityY += dy * this._mass * invDist3 * Body.DT;
        this._velocityZ += dz * this._mass * invDist3 * Body.DT;
    }

    public void PositionIntegration()
    {
        this._coordinateX = this._velocityX * Body.DT;
        this._coordinateY = this._velocityY * Body.DT;
        this._coordinateZ = this._velocityZ * Body.DT;
    }
}
